hbase-client/src/main/java/org/apache/hadoop/hbase/client/RpcRetryingCallerImpl.java

```diff
@@ -224,13 +224,12 @@ static Throwable translateException(Throwable t) throws DoNotRetryIOException {
     if (t instanceof ServiceException) {
       ServiceException se = (ServiceException)t;
       Throwable cause = se.getCause();
-      if (cause != null && cause instanceof DoNotRetryIOException) {
+      if (cause instanceof DoNotRetryIOException) {
         throw (DoNotRetryIOException)cause;
       }
       // Don't let ServiceException out; its rpc specific.
-      t = cause;
-      // t could be a RemoteException so go around again.
-      translateException(t);
+      // It also could be a RemoteException, so go around again.
+      t = translateException(cause);
```

Comment:
Ugh, I dislike this recursion but a) it is what it is, at this point; and b) don't want to impose a refactor for our mess on a third party